### PR TITLE
Adaptive Routing and Switching

### PR DESCRIPTION
**Pull request by JaiOCP**

### CODE REVIEW

**Comment by lguohan** [comment]
is there example to share? is there any design doc with this proposal?

**Comment by JaiOCP** [comment]
I am adding the design doc and also SAI pipeline changes.On Nov 17, 2022, at 11:47 AM, Guohan Lu ***@***.***> wrote:﻿
is there example to share? is there any design doc with this proposal?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

This electronic communication and the information and any files transmitted with it, or attached to it, are confidential and are intended solely for the use of the individual or entity to whom it is addressed and may contain information that is confidential, legally privileged, protected by privacy laws, or otherwise restricted from disclosure to anyone else. If you are not the intended recipient or the person responsible for delivering the e-mail to the intended recipient, you are hereby notified that any use, copying, distributing, dissemination, forwarding, printing, or copying of this e-mail is strictly prohibited. If you received this e-mail in error, please return the e-mail to the sender, delete it from your computer, and destroy any printed copy of it.

File: inc/saiacl.h [hunk 1]
```diff
@@ -277,6 +277,15 @@ typedef enum _sai_acl_action_type_t
     /** Set Forwarding class */
     SAI_ACL_ACTION_TYPE_SET_FORWARDING_CLASS = 0x00000034,
 
+    /** Set ARS monitoring */
+    SAI_ACL_ACTION_TYPE_SET_ARS_MONITORING = 0x00000035,
+
+    /** Set ARS object */
+    SAI_ACL_ACTION_TYPE_SET_ARS_OBJECT = 0x00000036,
```
Comment (eddyk-nvidia) [inline inc/saiacl.h:284]: Please clarify how setting the ARS Oid in ACL action relates to a ARS Object (single) bound to NHG object. As we discussed, multiple ARS objects (profiles) can be defined and used e.g. for TCP and non-TCP traffic. Still a single ARS object is on NHG object

File: inc/sainexthopgroup.h [hunk 2]
```diff
@@ -184,6 +184,41 @@ typedef enum _sai_next_hop_group_attr_t
...
+    /**
+     * @brief Adaptive routing and switching object for this group.
+     *
+     * @type sai_object_id_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_ARS
+     * @allownull true
+     * @default SAI_NULL_OBJECT_ID
+     */
+    SAI_NEXT_HOP_GROUP_ATTR_ARS_OBJECT_ID,
```
Comment (eddyk-nvidia) [inline inc/sainexthopgroup.h:196]: Probably the question for SONIC folks. What is the trigger for binding the ARS object to NHG ? How is the ARS object chosen (multiple objects/profiles) can be created

File: inc/sai.h [hunk 3]
```diff
@@ -140,6 +142,8 @@ typedef enum _sai_api_t
     SAI_API_MY_MAC           = 45, /**< sai_my_mac_api_t */
     SAI_API_IPSEC            = 46, /**< sai_ipsec_api_t */
     SAI_API_GENERIC_PROGRAMMABLE = 47, /**<sai_generic_programmable_t */
+    SAI_API_ARS              = 48, /**< sai_ars_api_t */
+    SAI_API_ARS_QUALITY_MAP  = 49, /**< sai_ars_quality_map_api_t */
     SAI_API_MAX,                   /**< total number of APIs */
 } sai_api_t;
```
Comment (eddyk-nvidia) [inline inc/sai.h:149]: In the previous commit there were additions to the Port Object (ARS enable and more). Removed from the commit ?

File: inc/saiarsqualitymap.h [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_ARS_RANDOM_SEED,
+
+    /**
+     * @brief Maximum number of ECMP ARS groups
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_ECMP_ARS_MAX_GROUPS,
```
Comment (rck-innovium) [inline inc/saiarsqualitymap.h:89]: In the current version of this proposal, several attributes of ARS_QUALITY_MAP do not logically belong to "Quality Map".

**Review by rck-innovium — CHANGES_REQUESTED** [review]
The SAI_ARS_QUALITY_MAP_ class has several attributes which are not per-se Quality map related.

Example:

- SAI_ARS_QUALITY_MAP_ATTR_ALGO
- SAI_ARS_QUALITY_MAP_ATTR_SAMPLING_INTERVAL
- SAI_ARS_QUALITY_MAP_ATTR_ARS_RANDOM_SEED
- SAI_ARS_QUALITY_MAP_ATTR_ECMP_ARS_MAX_GROUPS
- SAI_ARS_QUALITY_MAP_ATTR_ECMP_ARS_MAX_MEMBERS_PER_GROUP
- SAI_ARS_QUALITY_MAP_ATTR_LAG_ARS_MAX_MEMBERS_PER_GROUP

I presume that the intent of placing them here is to avoid adding too many ARS related attributes directly into the switch. However, it would be conceptually clearer to create few object types and attach instances of this object type to the switch. For example, something like the below:

ARS_PROFILE class to model configuration related attributes:
- SAI_ARS_PROFILE_ATTR_ALGO
- SAI_ARS_PROFILE_ATTR_ARS_RANDOM_SEED

ARS_CAPABILITY class to model capability related attributes:
- SAI_ARS_CAPABILITY_ATTR_ECMP_ARS_MAX_GROUPS
- SAI_ARS_CAPABILITY_ATTR_ECMP_ARS_MAX_MEMBERS_PER_GROUP
- SAI_ARS_CAPABILITY_ATTR_LAG_ARS_MAX_GROUPS

And add few switch attributes like:

- SAI_SWITCH_ATTR_ARS_PROFILE
- SAI_SWITCH_ATTR_ARS_CAPABILITY

File: inc/saiacl.h (hunk 1 above)
Comment (JaiOCP) [reply inc/saiacl.h:284]: Route lookup will result in a NHG with ARS object. This is correct. There is 1:1 relationship between NHG and ARS object.

Following two ways can be used to assign an ARS object for a specific flow type.
1. When a flow resolves via NHG with ARS object.
e.g. match<flow-type-tcp> action <SAI_ACL_ACTION_TYPE_DISABLE_ARS_FORWARDING>
In this case TCP flows resolving via NHG with ARS object will go through normal NHG object processing. This rule can be used as a catch all rule for all TCP traffic type, even if NHG do not have any ARS object present and will be noop in this case.

2. When a flow resolves via NHG without ARS object
e.g. match<flow-type-udp> action < SAI_ACL_ACTION_TYPE_SET_ARS_OBJECT>
In this case UDP flows will undergo ARS object processing.
Note: (2) can be used to override ARS object as present in NHG as well. In this case ARS object OID as provided by the ACL rule will be used instead of NHG provided ARS OID.

File: inc/sai.h (hunk 3 above)
Comment (JaiOCP) [reply inc/sai.h:149]: Added it back. Thanks.
https://github.com/opencomputeproject/SAI/pull/1658/commits/4b7f7da34af48d86e3d72840256f19def1a369d2

**Comment by JaiOCP** [comment]
> The SAI_ARS_QUALITY_MAP_ class has several attributes which are not per-se Quality map related.
> 
> Example:
> 
> * SAI_ARS_QUALITY_MAP_ATTR_ALGO
> * SAI_ARS_QUALITY_MAP_ATTR_SAMPLING_INTERVAL
> * SAI_ARS_QUALITY_MAP_ATTR_ARS_RANDOM_SEED
> * SAI_ARS_QUALITY_MAP_ATTR_ECMP_ARS_MAX_GROUPS
> * SAI_ARS_QUALITY_MAP_ATTR_ECMP_ARS_MAX_MEMBERS_PER_GROUP
> * SAI_ARS_QUALITY_MAP_ATTR_LAG_ARS_MAX_MEMBERS_PER_GROUP
> 
> I presume that the intent of placing them here is to avoid adding too many ARS related attributes directly into the switch. However, it would be conceptually clearer to create few object types and attach instances of this object type to the switch. For example, something like the below:
> 
> ARS_PROFILE class to model configuration related attributes:
> 
> * SAI_ARS_PROFILE_ATTR_ALGO
> * SAI_ARS_PROFILE_ATTR_ARS_RANDOM_SEED
> 
> ARS_CAPABILITY class to model capability related attributes:
> 
> * SAI_ARS_CAPABILITY_ATTR_ECMP_ARS_MAX_GROUPS
> * SAI_ARS_CAPABILITY_ATTR_ECMP_ARS_MAX_MEMBERS_PER_GROUP
> * SAI_ARS_CAPABILITY_ATTR_LAG_ARS_MAX_GROUPS
> 
> And add few switch attributes like:
> 
> * SAI_SWITCH_ATTR_ARS_PROFILE
> * SAI_SWITCH_ATTR_ARS_CAPABILITY

Hi Ravi,
Thanks for the review comments.
Lets discuss this in community meeting. I agree that these attributes really are not related to quality map but creating so many new objects just for compaction is also not a good idea. But lets discuss and see what others have to say.

File: inc/saiarsqualitymap.h (hunk 4 above)
Comment (JaiOCP) [reply inc/saiarsqualitymap.h:89]: Lets discuss in community meeting

File: inc/saiport.h [hunk 5]
```diff
@@ -2160,6 +2160,44 @@ typedef enum _sai_port_attr_t
...
     SAI_PORT_ATTR_MAX_FEC_SYMBOL_ERRORS_DETECTABLE,
 
+    /**
+     * @brief Enable ARS on port. This enables the control for quality measure in switch pipeline.
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_PORT_ATTR_ARS_ENABLE,
```
Comment (marian-pritsak) [inline inc/saiport.h:2170]: Can a port RIF with disabled ARS be part of the next hop group with ARS enabled? Or is this attribute only related to LAG?

File: inc/saiarsqualitymap.h (hunk 4 above)
Comment (JaiOCP) [reply inc/saiarsqualitymap.h:89]: Change ars quality map object name to ars profile object.

File: inc/saiport.h (hunk 5 above)
Comment (JaiOCP) [reply inc/saiport.h:2170]: Good practice is to enable all the ports for ARS participation and not worry about NHG resolution for a port that is not enabled for ARS.

Having said that if there is a an advantage to disable ARS for certain ports or port groups (like for power etc), and if NHG resolves via port not enabled for ARS, it will effectively not participate in the ARS group and will not get any traffic scheduled. This is easy to observe in a deployment as no traffic will be forwarded to the next hop pointing to this link.

File: inc/sai.h [hunk 6]
```diff
@@ -140,6 +142,8 @@ typedef enum _sai_api_t
     SAI_API_MY_MAC           = 45, /**< sai_my_mac_api_t */
     SAI_API_IPSEC            = 46, /**< sai_ipsec_api_t */
     SAI_API_GENERIC_PROGRAMMABLE = 47, /**<sai_generic_programmable_t */
+    SAI_API_ARS              = 48, /**< sai_ars_api_t */
```
Comment (ashutosh-agrawal) [inline inc/sai.h:145]: Was there a design document in the earlier version of this PR? If yes, can you please add it in this PR too.

Comment (ashutosh-agrawal) [inline inc/sai.h:145]: Are you planning to amend SAI pipeline diagram to indicate how the new tables/objects fit in the existing pipeline model?

Comment (JaiOCP) [reply inc/sai.h:145]: https://github.com/opencomputeproject/SAI/pull/1658/commits/05e9558fe4d467f3b0881d6e6ca3e57b0bd923ba

Comment (JaiOCP) [reply inc/sai.h:145]: Design Doc has been added with this commit

Comment (JaiOCP) [reply inc/sai.h:145]: This feature appears just another detail of ECMP, I feel it is not necessary.

File: inc/saiarsqualitymap.h (hunk 4 above)
Comment (JaiOCP) [reply inc/saiarsqualitymap.h]: Chnaged the  ARS quality map object to ARS profile object.
https://github.com/opencomputeproject/SAI/pull/1658/commits/05e9558fe4d467f3b0881d6e6ca3e57b0bd923ba

File: inc/sainexthopgroup.h (hunk 2 above)
Comment (JaiOCP) [reply inc/sainexthopgroup.h:196]: As discussed, this is part of SONiC HLD that is being worked on.

File: inc/sai.h (hunk 6 above)
Comment (ashutosh-agrawal) [reply inc/sai.h:145]: Even if don't update the existing pipeline diagrams, I think it would help to add a logical pipeline diagram in this PR itself.

**Review by rck-innovium — CHANGES_REQUESTED** [review]
Please include the nexthop group member attributes to model primary and alternate paths.

File: doc/SAI-Proposal-ARS.md [hunk 7]
```diff
@@ -0,0 +1,1327 @@
...
+      - *random path assignment for each packet*
+
+3. Fixed: fixed path assignment
+
+      - *mostly used for debugging*
+
+## 4.0 ARS Packet Flow
+
+![Adaptive Routing and Switching](../figures/ARS-Packet-Flow.png "Figure 1: ARS Packet Flow")
+
```
Comment (ashutosh-agrawal) [inline doc/SAI-Proposal-ARS.md:117]: This diagram probably needs a bit of a cleanup and some explanation. Are we storing micro-flows or macro-flows in this table? Also, is there a single logical instance of this table? Or there are n logical tables, one for each macro-flow. Also, it would help  if the color scheme used to represent a micro-flows is consistent throughout the diagram.

File: doc/ECMP/Hierarchical-ECMP.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+Hierarchical ECMP
```
Comment (ashutosh-agrawal) [inline doc/ECMP/Hierarchical-ECMP.md:1]: This MD file probably belongs to a different PR and was included here by mistake.

File: doc/SAI-Proposal-ARS.md [hunk 9]
```diff
@@ -0,0 +1,1327 @@
...
+| Flowlet    | contagious subsequence of packets within a macro flow separated by an idle time interval    |
+| Macro Flow Table    | HW table to store per macro flow state    |
+| EWMA | Exponential weighted moving average |
+| ARS    | Adaptive routing and switching: Path selection is influenced by quality metrics    |
+| NHG | Next Hop Group |
+
+### 3.2 Key Idea behind ARS
+1. ARS is flowlet based
+
+      - *A flow is viewed as a sequence of flowlets separated by idle time*
```
Comment (ashutosh-agrawal) [inline doc/SAI-Proposal-ARS.md:61]: Nit: Should the "flow" in this section be replaced by "Macro Flow"? Traditionally, we think of micro-flows as flows, so explicit usage of micro-flow or macro-flow in the rest of this document will help avoid misunderstanding.

File: doc/SAI-Proposal-ARS.md [hunk 10]
```diff
@@ -0,0 +1,1327 @@
...
+
+      - *ARS can be used to partition a NHG into a primary and alternate set of paths. Alternate paths are typically non minimal cost or side links based on the topology. The alternate paths are chosen when primary paths are relatively worse.*
+
+7. LAG
+
+      - *Besides being able to choose best quality path from the NHG, packet should also be able to select best link available amongst LAG memembers*
+
+8. Link Failure or Admin Down
+
+      - *Besides being able to choose best quality path from the NHG, packet should also be able to select best link available amongst LAG memembers*
```
Comment (ashutosh-agrawal) [inline doc/SAI-Proposal-ARS.md:92]: Copy-paste of the previous bullet on LAG.

File: inc/sai.h (hunk 6 above)
Comment (JaiOCP) [reply inc/sai.h:145]: Added a diagram in the new PR

File: inc/saiarsprofile.h [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+
+    /**
+     * @brief Enable past port load as the quality parameter. This is the average egress bytes measured on a port
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
+     * @validonly SAI_ARS_PROFILE_ATTR_ALGO == SAI_ARS_PROFILE_ALGO_EWMA
+     */
+    SAI_ARS_PROFILE_ATTR_PORT_LOAD_PAST,
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply inc/saiarsprofile.h:123]: This is granular control. I though about removing but there might be a usecase to disable participating link for load computation so keeping it.

File: inc/sainexthopgroup.h [hunk 12]
```diff
@@ -327,6 +362,33 @@ typedef enum _sai_next_hop_group_member_attr_t
...
+    SAI_NEXT_HOP_GROUP_MEMBER_ATTR_ARS_ALTERNATE_PATH_COST,
+
+    /**
+     * @brief Indicates the bias in favor of alternate path
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_NEXT_HOP_GROUP_MEMBER_ATTR_ARS_ALTERNATE_PATH_BIAS,
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply inc/sainexthopgroup.h:390]: Bias is a probability of selecting for randomization.

File: inc/saiars.h [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    SAI_ARS_ATTR_SAMPLEPACKET_ENABLE,
+
+    /**
+     * @brief Maximum number of alternate members per adaptive routing group
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 16
+     */
+    SAI_ARS_ATTR_MAX_ALT_MEMEBERS_PER_GROUP,
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply inc/saiars.h:125]: > How does this work? If the same ARS object is associated with two different NHGs, does it mean that the NHGs should have the same number of primary and secondary members?

yes. only if there is common adaptive behavior NHG should share the same ARS object.

File: doc/SAI-Proposal-ARS.md (hunk 7 above)
Comment (JaiOCP) [reply doc/SAI-Proposal-ARS.md:117]: Thanks for this. Cleaned it up in the new PR.

File: doc/SAI-Proposal-ARS.md (hunk 9 above)
Comment (JaiOCP) [reply doc/SAI-Proposal-ARS.md:61]: Flow and Macro flow is used interchangeably and given context should help. I have opened a new PR and would avoid churn in that if possible.

**Comment by JaiOCP** [comment]
Closing this PR as enums have been shifted and can't be squashed.
https://github.com/opencomputeproject/SAI/pull/1681
